# Read database records into pandas dataFrame and write it back

This will cover:
- pd.read_sql_table
- pd.read_sql_query
- df.to_sql
- pd.read_sql (which is just a wrapper for the first two methods)

## Create the table by using PGAdmin

The reason I am using posgreSql because I am familiar with it since I mostly work with PostgreSQL when working in Django and FastAPI. You are Free to use any database of your choose the steps are the same only some of the implementation will change if you choose to use a non-relational database example : MongoDB. 
Libraries that I had to install : 
1. psycopg
2. SQLAlchemy (manage database connections)
3. python-decouple (To get the database creds from .env file)
4. Pydantic (You can skip it if you don't want to write type safe functions) 
<br>

```
-- Create a customers table
CREATE TABLE Customers (
id BIGSERIAL PRIMARY KEY,
name VARCHAR(255) NOT NULL,
phone_number VARCHAR(10) 
);

-- Insert data into the customers table
INSERT INTO Customers (name,phone_number) VALUES
('Rollex', '9972688590'),
('Ballistic', '1000688590'),
('Barricade', '2472688590'),
('Bullet', '7622688590'),
('Barbatos', '1172688590'),
('Artillery', '1502688590'),
('Shotgun', '1852688590'),
('Kombat', '3652688590'),
('Lockdown', '1442688590'),
('Ignition', '6000688590'),
('Shockwave', '3602688590'),
('PointBreak', '7472688590'),
('GroundZero', '8002688590'),
('Interceptor','9112688590');

-- check if the data has been successfully inserted into the customers table
SELECT * FROM Customers;

-- Create an orders table
CREATE TABLE Orders (
id BIGSERIAL PRIMARY KEY,
customer_id BIGINT NOT NULL,
name VARCHAR(255) NOT NULL,
amount MONEY NOT NULL,
order_date TIMESTAMP DEFAULT NOW(),
CONSTRAINT fk_customer FOREIGN KEY (customer_id)
REFERENCES Customers (id)
ON DELETE CASCADE
);

-- insert the data into the orders table
INSERT INTO Orders (customer_id,name,amount) VALUES
(1,'50 calibure fully automatic Browning machine gun',9978.123),
(2,'50 calibure desert eagle handgun',1000.123),
(3,'500 S and W Magnum',2470.123),
(4,'Anzio 20mm anti material sniper rifle',7620.123),
(5,'Beretta m9 9mm',1170.123),
(6,'Milkor MGL (40mm Multiple Grenade Launcher)',1500.123),
(7,'Benelli M4 (M1014) 12 gauge shotgun shell',1850.123),
(8,'AK-47',3650.123),
(9,'M249 SAW Mk 48 Mod 1 7.62×51mm',3650.123),
(10,'Javelin rocket launcher',6000.123),
(11,'Milkor MGL (40mm Multiple Grenade Launcher)',3600.123),
(12,'FN SCAR-H (Special Operations Forces Combat Assault Rifle - Heavy)',7470.123),
(13,'M249 SAW Mk 48 Mod 1 7.62×51mm',8000.123),
(14,'Beretta m9 9mm',9110.123);

-- check if the data is present in the orders table
SELECT * FROM Orders;
```

## Import all the necessary packages

In [16]:
from pydantic import BaseModel, Field, ValidationError
from typing import List,Optional
from decouple import config, Config,RepositoryEnv

from sqlalchemy import create_engine
from sqlalchemy.engine import Engine
from sqlalchemy import inspect
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import itertools

# import path so that I can tell pandas where to look for env file in the file system.
from pathlib import Path

## Create a Pydantic model for database configuration

In [17]:
class DBConfig(BaseModel):
    db_name:str=Field(default=None)
    user:str=Field(default=None)
    password:str=Field(default=None)
    host:str=Field(default="localhost")
    port:str=Field(default=5432)

## Create a common function to create a database connection using sqlAlchemy database engine

In [18]:
def database_engine(config:DBConfig) -> Engine:
    engine = create_engine(
        f"postgresql+psycopg2://{config.user}:{config.password}@{config.host}:{config.port}/{config.db_name}"
    )
    return engine

## Tell pandas where to look for env file

In [19]:

env_path = Path("/home/aditya/github/Deep-learning-prerequisite/pandas/pandas_code/database_to_pandas/.env")
config = Config(RepositoryEnv(env_path))

## Change the working directory in Jupyter

In [20]:
import os
os.chdir("/home/aditya/github/Deep-learning-prerequisite/pandas/pandas_code/database_to_pandas/")

## Create db engine object

In [21]:
db_config=DBConfig(
    db_name=config("DB_NAME"),
    user=config("DB_USER"),
    password=config("DB_PASSWORD"),
    host=config("DB_HOST"),
    port=config("DB_PORT")
)
engine_obj=database_engine(db_config)
engine_obj

Engine(postgresql+psycopg2://postgres:***@localhost:5432/Pandas)

## Create a Pydantic model for reading table from database

In [22]:
class ReadDbTableModel(BaseModel):
    table_name:str=Field(default=None)
    engine:Engine

PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'sqlalchemy.engine.base.Engine'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.11/u/schema-for-unknown-type

### Analysis : 
```
PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'sqlalchemy.engine.base.Engine'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
```

<br>

This error is acutally coming from Pydantic and it's expected when I try to use arbitrary Python types (like ```sqlalchemy.engine.base.Engine```) as fields.
By default Pydantic cannot generate a schema for ```Engine``` because it's not a "standard" type, You have two ways to fix it:
### Solution:
There are two solutions to this predicament that I am in 
- Solution 1 : Use a validator or store as Any
  - If you don't care about the type checking for engine but just want Pydantic to accept it . This works but you lose "type hint" benefits.
- Solution 2 : Allow arbitary types
  - You can tell Pydantic to allow any type in the model using ```model_config```:

<br>

I went with solution 2 because I don't want to loose type hint.

In [23]:
class ReadDbTableModel(BaseModel):
    table_name:str=Field(default=None)
    columns:List[str]=Field(default=[])
    engine:Engine
    model_config = {
    "arbitrary_types_allowed":True
    }

In [24]:
def read_db_table(config:ReadDbTableModel) -> pd.DataFrame:
    if not config.table_name:
        raise ValueError("Table name is required!")
    try:
        if not len(config.columns): 
            df = pd.read_sql_table(config.table_name,con=config.engine)
        else:
            df = pd.read_sql_table(config.table_name, con=config.engine, columns=config.columns)
        return df
    except Exception as e:
        raise RuntimeError(f"Failed to read table '{config.table_name}': {e}")

### NOTICE : 
The reason I don't need to manually open or close connection here:
- ```pandas.read_sql_table()``` uses the SQLALchemy engine (```engine```)
- ```SQLAlchemy``` engines manage connection pooling automatically. When I call ```read_sql_table```:
  - It borrows a connection from the pool.
  - Executes the query.
  - Closes (releases) the connection back to the pool when done.
- So connections are managed automatically.

<br>

If I were using a raw connection (like ```psycopg2.connect()```), I'd have to open/close manually. But since I am using ```SQLAlchemy engine``` + ```pandas```, I don't need to.

## Generate a dataFrame from the sql_table from the database

### Get customers data from the database

In [25]:
config_read_customers_table = ReadDbTableModel(table_name="customers", engine=engine_obj)
customers_df = read_db_table(config_read_customers_table)
customers_df

,id,name,phone_number
0,1,Rollex,9972688590
1,2,Ballistic,1000688590
2,3,Barricade,2472688590
3,4,Bullet,7622688590
4,5,Barbatos,1172688590
5,6,Artillery,1502688590
6,7,Shotgun,1852688590
7,8,Kombat,3652688590
8,9,Lockdown,1442688590
9,10,Ignition,6000688590


### Get customers data from database and create dataFrame of the selected columns

#### Get data and create dataFrame of selected name column

In [26]:
config_read_customers_name_table = ReadDbTableModel(table_name="customers", engine=engine_obj,columns=["name"])
customers_name_df = read_db_table(config_read_customers_name_table)
customers_name_df

,name
0,Rollex
1,Ballistic
2,Barricade
3,Bullet
4,Barbatos
5,Artillery
6,Shotgun
7,Kombat
8,Lockdown
9,Ignition


#### Get data and create dataFrame of selected phone_number column

In [27]:
config_read_customers_phone_no_table = ReadDbTableModel(table_name="customers",engine=engine_obj,columns=["phone_number"])
customer_phone_number_df = read_db_table(config_read_customers_phone_no_table)
customer_phone_number_df

,phone_number
0,9972688590
1,1000688590
2,2472688590
3,7622688590
4,1172688590
5,1502688590
6,1852688590
7,3652688590
8,1442688590
9,6000688590


### Get orders data from the database

In [30]:
config_read_orders_table = ReadDbTableModel(table_name="orders", engine=engine_obj)
orders_df = read_db_table(config_read_orders_table)
orders_df

,id,customer_id,name,amount,order_date
0,1,1,50 calibure fully automatic Browning machine gun,"$9,978.12",2025-09-03 14:09:23.202945
1,2,2,50 calibure desert eagle handgun,"$1,000.12",2025-09-03 14:09:23.202945
2,3,3,500 S and W Magnum,"$2,470.12",2025-09-03 14:09:23.202945
3,4,4,Anzio 20mm anti material sniper rifle,"$7,620.12",2025-09-03 14:09:23.202945
4,5,5,Beretta m9 9mm,"$1,170.12",2025-09-03 14:09:23.202945
5,6,6,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12",2025-09-03 14:09:23.202945
6,7,7,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12",2025-09-03 14:09:23.202945
7,8,8,AK-47,"$3,650.12",2025-09-03 14:09:23.202945
8,9,9,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12",2025-09-03 14:09:23.202945
9,10,10,Javelin rocket launcher,"$6,000.12",2025-09-03 14:09:23.202945


#### Read data from the database and only create dataFrame of selected name and amount columns

In [31]:
config_read_orders_table = ReadDbTableModel(table_name="orders",engine=engine_obj,columns=["name","amount"])
oders_name_amount_df = read_db_table(config_read_orders_table)
oders_name_amount_df

,name,amount
0,50 calibure fully automatic Browning machine gun,"$9,978.12"
1,50 calibure desert eagle handgun,"$1,000.12"
2,500 S and W Magnum,"$2,470.12"
3,Anzio 20mm anti material sniper rifle,"$7,620.12"
4,Beretta m9 9mm,"$1,170.12"
5,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12"
6,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12"
7,AK-47,"$3,650.12"
8,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12"
9,Javelin rocket launcher,"$6,000.12"


#### Read data from the database and only create dataFrame of selected name and order_date columns

In [32]:
config_read_orders_table = ReadDbTableModel(table_name="orders",engine=engine_obj,columns=["name","order_date"])
oders_name_order_date_df = read_db_table(config_read_orders_table)
oders_name_order_date_df

,name,order_date
0,50 calibure fully automatic Browning machine gun,2025-09-03 14:09:23.202945
1,50 calibure desert eagle handgun,2025-09-03 14:09:23.202945
2,500 S and W Magnum,2025-09-03 14:09:23.202945
3,Anzio 20mm anti material sniper rifle,2025-09-03 14:09:23.202945
4,Beretta m9 9mm,2025-09-03 14:09:23.202945
5,Milkor MGL (40mm Multiple Grenade Launcher),2025-09-03 14:09:23.202945
6,Benelli M4 (M1014) 12 gauge shotgun shell,2025-09-03 14:09:23.202945
7,AK-47,2025-09-03 14:09:23.202945
8,M249 SAW Mk 48 Mod 1 7.62×51mm,2025-09-03 14:09:23.202945
9,Javelin rocket launcher,2025-09-03 14:09:23.202945


## There is another way to create a dataFrame by reading table from the database that is using raw SQL Queries in Pandas

In [33]:
query = """SELECT 
                c.id AS customer_id , 
                c.name AS customer_name,
                c.phone_number as phone_no,
                o.id AS order_id,
                o.name AS order_name,
                o.amount,
                o.order_date
            FROM customers c
            INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC;
        """
df=pd.read_sql_query(query,engine_obj)
df

,customer_id,customer_name,phone_no,order_id,order_name,amount,order_date
0,1,Rollex,9972688590,1,50 calibure fully automatic Browning machine gun,"$9,978.12",2025-09-03 14:09:23.202945
1,2,Ballistic,1000688590,2,50 calibure desert eagle handgun,"$1,000.12",2025-09-03 14:09:23.202945
2,3,Barricade,2472688590,3,500 S and W Magnum,"$2,470.12",2025-09-03 14:09:23.202945
3,4,Bullet,7622688590,4,Anzio 20mm anti material sniper rifle,"$7,620.12",2025-09-03 14:09:23.202945
4,5,Barbatos,1172688590,5,Beretta m9 9mm,"$1,170.12",2025-09-03 14:09:23.202945
5,6,Artillery,1502688590,6,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12",2025-09-03 14:09:23.202945
6,7,Shotgun,1852688590,7,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12",2025-09-03 14:09:23.202945
7,8,Kombat,3652688590,8,AK-47,"$3,650.12",2025-09-03 14:09:23.202945
8,9,Lockdown,1442688590,9,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12",2025-09-03 14:09:23.202945
9,10,Ignition,6000688590,10,Javelin rocket launcher,"$6,000.12",2025-09-03 14:09:23.202945


### Analysis
- Now this works fine but according to what I have learned from my past experience that raw SQL queries must not exist in the code (correct me if I am wrong). 
- In django there is a way that you can use raw sql queries instead of built in ORM. In some cases when executing complex queries we may have to use raw sql queries. It is very tempting but it must be avoided at all cost.
- For example : Lets's suppose there are two tables customers and orders and you want to join the two table and return the data to the front-end. It is very tempting to just use raw sql query like this
  - Create a serializer : 
  ```python

    from rest_framework import serializers
    
    class CustomerOrderSerializer(serializers.Serializer):
        customer_id = serializers.IntegerField()
        customer_name = serializers.CharField()
        phone_number = serializers.CharField()
        order_id = serializers.IntegerField()
        order_name = serializers.CharField()
        amount = serializers.DecimalField(max_digits=12, decimal_places=2)
        order_date = serializers.DateTimeField()
  ```
   - This serialzies the data into JSON from a python object.
  - Create a view:
    ```python
    from django.db import connection
    from rest_framework.views import APIView
    from rest_framework.response import Response
    from .serializers import CustomerOrderSerializer
    
    class CustomerOrderListView(APIView):
        def get(self, request):
            query = """
            SELECT 
                c.id AS customer_id,
                c.name AS customer_name,
                c.phone_number,
                o.id AS order_id,
                o.name AS order_name,
                o.amount,
                o.order_date
            FROM Customers c
            INNER JOIN Orders o ON c.id = o.customer_id
            ORDER BY c.id;
            """
            with connection.cursor() as cursor:
                cursor.execute(query)
                columns = [col[0] for col in cursor.description]
                results = [
                    dict(zip(columns, row))
                    for row in cursor.fetchall()
                ]
    
            serializer = CustomerOrderSerializer(results, many=True)
            return Response(serializer.data)    
    ```
    - As you can see in this example I have used a raw SQL query to join the two table togeather and then sending it to the front-end (BEWARE don't do it). I literally got chewed on by my senior during my PR review when I was just a traniee for this unforgivable blasphemy.
### Reasons why raw sql queries must be avoided?
- For basic CRUD operations (Create, Read, Update, Delete) – the ORM is safer, cleaner, and prevents SQL injection.
- For maintainability – ORM queries adapt better if you change your database engine (e.g., PostgreSQL → MySQL).
- For security – ORM parameterization reduces risk of SQL injection.
### So what to do if raw sql queries must be avoided in code and you still want to execute complex sql query
- One of the way to do that is use a database Views or database functions depending on what you are trying to do.
- Create a model
  ```python
    class CustomerOrder(models.Model):
    customer_id = models.BigIntegerField()
    customer_name = models.CharField(max_length=255)
    phone_number = models.CharField(max_length=10)
    order_id = models.BigIntegerField()
    order_name = models.CharField(max_length=255)
    amount = models.DecimalField(max_digits=12, decimal_places=2)
    order_date = models.DateTimeField()

    class Meta:
        managed = False  # Django won't try to create or alter this table
        db_table = "customer_orders_view"
  ```
  Now I can use Django ORM queries <br>
  ```python
      CustomerOrder.objects.filter(customer_name__icontains="Rollex").order_by("-order_date")
  ```
### Reasons to use Database Views instead of raw sql queries 
- Encaplusation of complexity:
  - I defined the ```INNER JOIN``` (or any complex query) once in the database.
  - In Django, I can query the view like a normal table -- no raw sql query needed every time.
- Maintainability
  - If my query logic changes, I just need to update the view in PostgreSQL.
  - Django code does not need to change, as long as the column names in the view remain the same.
- Security
  - No direct raw SQL in Python; ORM queries against the view are automatically parameterized.
  - Reduces risk of SQL injection.

## Let's use database Views to fetch the data from the database in Pandas 
For the reasons which I discussed above by taking the example of django. I am going to try to achieve the similar results i.e create a dataFrame but instead of executing raw sql query I am going to use database View. <br>
Now before I start working on Pandas I need to create a database View that I can use in here <br>
```sql
    CREATE OR REPLACE FUNCTION get_customer_orders()
    RETURNS TABLE (
    customer_id BIGINT,
    customer_name VARCHAR,
    phone_number VARCHAR,
    order_id BIGINT,
    order_name VARCHAR,
    amount MONEY,
    order_date TIMESTAMP
    )
    LANGUAGE plpgsql
    AS $$
    BEGIN
    
    	RETURN QUERY
    	SELECT 
    		c.id AS customer_id,
    		c.name AS customer_name,
    		c.phone_number,
    		o.id AS order_id,
    		o.name AS order_name,
    		o.amount,
    		o.order_date
    	FROM customers c 
    	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC;
    END;
    $$
    
    CREATE OR REPLACE VIEW customer_orders_view AS
    SELECT * FROM get_customer_orders();
```
<br>
Test if the view created works or not : <br>

```sql
    SELECT * FROM customer_orders_view;
```

### I am going to create a common function that reads the data from the database and returns a dataFrame which can be used everywhere

#### Create a pydantic model to make the function type safe

In [34]:
class ReadDBViewModel(BaseModel):
    database_view_name:str=Field(default=None)
    table_name:str=Field(default=None)
    columns:List[str]=Field(default=[])
    engine:Engine
    model_config={
        "arbitrary_types_allowed":True
    }

In [35]:
def read_db_table_db_view_support(config:ReadDBViewModel) -> pd.DataFrame:
    if not config.table_name and not config.database_view_name:
        raise ValueError("Either provide table_name or database_view_name to successfully generate the dataFrame!")
    if config.database_view_name and not config.table_name:
        try:
            if not len(config.columns):
                df = pd.read_sql_table(f"SELECT * FROM {config.database_view_name}",con=config.engine)
            else:
                df = pd.read_sql_table(query,con=config.engine,columns=config.columns)
            return df
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.database_view_name}' : {e}")
    if config.table_name and not config.database_view_name:
        try:
            if not len(config.columns):
                df = pd.read_sql_table(config.table_name,con=config.engine)
            else:
                df = pd.read_sql_table(config.table_name,con=config.engine,columns=config.columns)
            return df
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}' : {e}")

In [36]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

RuntimeError: Failed to read table 'customer_orders_view' : Table SELECT * FROM customer_orders_view not found

### Analysis:
```pd.read_sql_table()``` <br>
- This only works with actual tables in the database (or SQLAlchemy table objects)
- You cannot pass a full SQL query like
  ```python
        query = "SELECT * FROM customer_orders_view"
        df = pd.read_sql_table(query, con=engine_obj)  # ❌ fails
  ```
- Here, read_sql_table tries to find a table literally named ```"SELECT * FROM customer_orders_view"```, which does not exist, hence the error.
### Solution:
After some digging I found out that there is another method that could be used to get the dataFrame using the database View <br>
I have to use this function ```.read_sql_query()``` instead of this function ```read_sql_table()``` 
- ```.read_sql_table()``` function will look for table literally hence it was failing when I tried executing the database view using this query ```SELECT * FROM customer_orders_view```
- ```.read_sql_query()``` function will help us execute the query that will trigger our database View and allow us to get the dataFrame that we desire.

### read_db_table_db_view_support function version 1 : Fixed function that allows us to create dataFrames from database by either supplying table or database view in it as an argument. 

In [37]:
def read_db_table_db_view_support(config:ReadDBViewModel) -> pd.DataFrame:
    if not config.table_name and not config.database_view_name:
        raise ValueError("Either provide table_name or database_view_name to successfully generate the dataFrame!")
    if config.database_view_name and not config.table_name:
        try:
            if not len(config.columns):
                df = pd.read_sql_query(f"SELECT * FROM {config.database_view_name}",con=config.engine)
            else:
                df = pd.read_sql_query(f"SELECT * FROM {config.database_view_name}",con=config.engine,columns=config.columns)
            return df
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.database_view_name}' : {e}")
    if config.table_name and not config.database_view_name:
        try:
            if not len(config.columns):
                df = pd.read_sql_table(config.table_name,con=config.engine)
            else:
                df = pd.read_sql_table(config.table_name,con=config.engine,columns=config.columns)
            return df
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}' : {e}")

#### Create a dataFrame using database view 

In [38]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,customer_id,customer_name,phone_number,order_id,order_name,amount,order_date
0,1,Rollex,9972688590,1,50 calibure fully automatic Browning machine gun,"$9,978.12",2025-09-03 14:09:23.202945
1,2,Ballistic,1000688590,2,50 calibure desert eagle handgun,"$1,000.12",2025-09-03 14:09:23.202945
2,3,Barricade,2472688590,3,500 S and W Magnum,"$2,470.12",2025-09-03 14:09:23.202945
3,4,Bullet,7622688590,4,Anzio 20mm anti material sniper rifle,"$7,620.12",2025-09-03 14:09:23.202945
4,5,Barbatos,1172688590,5,Beretta m9 9mm,"$1,170.12",2025-09-03 14:09:23.202945
5,6,Artillery,1502688590,6,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12",2025-09-03 14:09:23.202945
6,7,Shotgun,1852688590,7,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12",2025-09-03 14:09:23.202945
7,8,Kombat,3652688590,8,AK-47,"$3,650.12",2025-09-03 14:09:23.202945
8,9,Lockdown,1442688590,9,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12",2025-09-03 14:09:23.202945
9,10,Ignition,6000688590,10,Javelin rocket launcher,"$6,000.12",2025-09-03 14:09:23.202945


#### Create a dataFrame using database View where customer_name, order_name is selected

In [39]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj,columns=["customer_name","order_name"])
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

RuntimeError: Failed to read table 'customer_orders_view' : read_sql_query() got an unexpected keyword argument 'columns'

### Analysis:
Huh? ... Intreseting ... It looks like,  unlike ```.read_sql_table()``` ```.read_sql_query()``` doesn't have a columns argument which means we cannot use it to select columns. <br>
You can refer to this doc if you like https://pandas.pydata.org/docs/reference/api/pandas.read_sql_query.html 
### Solution:
I have to manually modify the query itself in other words I need to make some changes in the function to make ```read_sql_query()``` function behave like ```.read_sql_table()``` function <br>
### Things I want to achieve:
- Write a function so that the end user who is calling that function can't tell the difference between ```.read_sql_table()``` and ```.read_sql_query()``` which means I will have to keep the arguments passed in this function the same

### read_db_table_db_view_support function version 2 : Adding the support for column selection when using database View to generate the dataFrame

In [40]:
def read_db_table_db_view_support(config:ReadDBViewModel) -> pd.DataFrame:
    if not config.table_name and not config.database_view_name:
        raise ValueError("Either provide table_name or database_view_name to successfully generate the dataFrame!")
    if config.database_view_name and not config.table_name:
        try:
            # database_view = 
            if not len(config.columns):
                df = pd.read_sql_query(f"SELECT * FROM {config.database_view_name}",con=config.engine)
            else:
                df = pd.read_sql_query(f"SELECT {', '.join(config.columns)} FROM {config.database_view_name}",con=config.engine)
            return df
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.database_view_name}' : {e}")
    if config.table_name and not config.database_view_name:
        try:
            if not len(config.columns):
                df = pd.read_sql_table(config.table_name,con=config.engine)
            else:
                df = pd.read_sql_table(config.table_name,con=config.engine,columns=config.columns)
            return df
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}' : {e}")

#### Generate a dataFrame from the database using database View using all the columns

In [41]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,customer_id,customer_name,phone_number,order_id,order_name,amount,order_date
0,1,Rollex,9972688590,1,50 calibure fully automatic Browning machine gun,"$9,978.12",2025-09-03 14:09:23.202945
1,2,Ballistic,1000688590,2,50 calibure desert eagle handgun,"$1,000.12",2025-09-03 14:09:23.202945
2,3,Barricade,2472688590,3,500 S and W Magnum,"$2,470.12",2025-09-03 14:09:23.202945
3,4,Bullet,7622688590,4,Anzio 20mm anti material sniper rifle,"$7,620.12",2025-09-03 14:09:23.202945
4,5,Barbatos,1172688590,5,Beretta m9 9mm,"$1,170.12",2025-09-03 14:09:23.202945
5,6,Artillery,1502688590,6,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12",2025-09-03 14:09:23.202945
6,7,Shotgun,1852688590,7,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12",2025-09-03 14:09:23.202945
7,8,Kombat,3652688590,8,AK-47,"$3,650.12",2025-09-03 14:09:23.202945
8,9,Lockdown,1442688590,9,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12",2025-09-03 14:09:23.202945
9,10,Ignition,6000688590,10,Javelin rocket launcher,"$6,000.12",2025-09-03 14:09:23.202945


#### Generate a dataFrame from database by using database Views after selecting the columns customer_name, order_name

In [42]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj,columns=["customer_name","order_name"])
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,customer_name,order_name
0,Rollex,50 calibure fully automatic Browning machine gun
1,Ballistic,50 calibure desert eagle handgun
2,Barricade,500 S and W Magnum
3,Bullet,Anzio 20mm anti material sniper rifle
4,Barbatos,Beretta m9 9mm
5,Artillery,Milkor MGL (40mm Multiple Grenade Launcher)
6,Shotgun,Benelli M4 (M1014) 12 gauge shotgun shell
7,Kombat,AK-47
8,Lockdown,M249 SAW Mk 48 Mod 1 7.62×51mm
9,Ignition,Javelin rocket launcher


#### Generate a dataFrame from database by using database Views after selecting the columns customer_name, order_name, amount

In [43]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj,columns=["customer_name","order_name","amount"])
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,customer_name,order_name,amount
0,Rollex,50 calibure fully automatic Browning machine gun,"$9,978.12"
1,Ballistic,50 calibure desert eagle handgun,"$1,000.12"
2,Barricade,500 S and W Magnum,"$2,470.12"
3,Bullet,Anzio 20mm anti material sniper rifle,"$7,620.12"
4,Barbatos,Beretta m9 9mm,"$1,170.12"
5,Artillery,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12"
6,Shotgun,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12"
7,Kombat,AK-47,"$3,650.12"
8,Lockdown,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12"
9,Ignition,Javelin rocket launcher,"$6,000.12"


## Now I am going to attempt to wirte the data from a csv file into the database

In [44]:
csv_orders_df=pd.read_csv("/home/aditya/github/Deep-learning-prerequisite/pandas/dataset/read_orders_table_read_db_df.csv")
csv_orders_df

,id,customer_id,name,amount,order_date
0,1,1,Axe,"$9,978.12",2025-09-01 12:01:26.625465
1,2,2,Sledge Hammer,"$1,000.12",2025-09-01 12:01:26.625465
2,3,3,Showel,"$2,470.12",2025-09-01 12:01:26.625465
3,4,4,Cross blade Axe,"$7,620.12",2025-09-01 12:01:26.625465
4,5,5,Pick Axe,"$1,170.12",2025-09-01 12:01:26.625465
5,6,6,Michetti,"$1,500.12",2025-09-01 12:01:26.625465
6,7,7,Crow-bar,"$1,850.12",2025-09-01 12:01:26.625465
7,8,8,Titanium base ball bat,"$3,650.12",2025-09-01 12:01:26.625465
8,9,9,Mace,"$3,650.12",2025-09-01 12:01:26.625465
9,10,10,Gurkha dagger,"$6,000.12",2025-09-01 12:01:26.625465


### In a same way that I had previously created a function for reading the database In the same way I want to write a function to write the data into the database after reading the data from a csv file

In [45]:
class WritePandasToDBModel(BaseModel):
    df:pd.DataFrame=Field(default=None)
    table_name:str=Field(default=None)
    index:bool=Field(default=False)
    if_exists:str=Field(default=False)
    engine:Engine
    model_config={
        "arbitrary_types_allowed":True
    }
def write_df_to_database(config:WritePandasToDBModel):
    if config.df is None or config.df.empty:
        raise ValueError("DataFrame is required and cannot be empty to write into the database")
    if not table_name:
        raise ValueError("Table name is required to be able to write the data into the database.")
    if config.table_name:
        try:
            # remove the id (primary key column) if exists
            if "id" in config.df.columns:
                config.df = config.df.drop(columns=["id"])
            config.df.to_sql(name=config.table_name,con=config.engine,index=config.index,if_exists=config.if_exists)
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}':{e}")

In [46]:
table_name = "orders"
engine=engine_obj
if_exists="append"
index=False
dataFrame=csv_orders_df
write_to_db = WritePandasToDBModel(df=dataFrame,table_name=table_name,engine=engine,if_exists=if_exists,index=index)
write_df_to_database =write_df_to_database(write_to_db)

#### To verify if it worked or not I am going to create another dataFrame using the orders table from the database

In [47]:
db_view_name = "orders"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,id,customer_id,name,amount,order_date
0,1,1,50 calibure fully automatic Browning machine gun,"$9,978.12",2025-09-03 14:09:23.202945
1,2,2,50 calibure desert eagle handgun,"$1,000.12",2025-09-03 14:09:23.202945
2,3,3,500 S and W Magnum,"$2,470.12",2025-09-03 14:09:23.202945
3,4,4,Anzio 20mm anti material sniper rifle,"$7,620.12",2025-09-03 14:09:23.202945
4,5,5,Beretta m9 9mm,"$1,170.12",2025-09-03 14:09:23.202945
5,6,6,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12",2025-09-03 14:09:23.202945
6,7,7,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12",2025-09-03 14:09:23.202945
7,8,8,AK-47,"$3,650.12",2025-09-03 14:09:23.202945
8,9,9,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12",2025-09-03 14:09:23.202945
9,10,10,Javelin rocket launcher,"$6,000.12",2025-09-03 14:09:23.202945


#### As you can see that the data in the database has been updated now there exists new rows of data

#### Let's see what is the output dataFrame when I trigger the database View that returns a table using inner join between customers and orders table

In [48]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,customer_id,customer_name,phone_number,order_id,order_name,amount,order_date
0,1,Rollex,9972688590,1,50 calibure fully automatic Browning machine gun,"$9,978.12",2025-09-03 14:09:23.202945
1,2,Ballistic,1000688590,2,50 calibure desert eagle handgun,"$1,000.12",2025-09-03 14:09:23.202945
2,3,Barricade,2472688590,3,500 S and W Magnum,"$2,470.12",2025-09-03 14:09:23.202945
3,4,Bullet,7622688590,4,Anzio 20mm anti material sniper rifle,"$7,620.12",2025-09-03 14:09:23.202945
4,5,Barbatos,1172688590,5,Beretta m9 9mm,"$1,170.12",2025-09-03 14:09:23.202945
5,6,Artillery,1502688590,6,Milkor MGL (40mm Multiple Grenade Launcher),"$1,500.12",2025-09-03 14:09:23.202945
6,7,Shotgun,1852688590,7,Benelli M4 (M1014) 12 gauge shotgun shell,"$1,850.12",2025-09-03 14:09:23.202945
7,8,Kombat,3652688590,8,AK-47,"$3,650.12",2025-09-03 14:09:23.202945
8,9,Lockdown,1442688590,9,M249 SAW Mk 48 Mod 1 7.62×51mm,"$3,650.12",2025-09-03 14:09:23.202945
9,10,Ignition,6000688590,10,Javelin rocket launcher,"$6,000.12",2025-09-03 14:09:23.202945


### Let's see what is the output dataFrame when I trigger the database View that returns a table using inner join between customers and orders table but there is a twist I am going to drop the orders table using PGAdmin and then use this function to save the data in the dataFrame into the database table.

In [49]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

RuntimeError: Failed to read table 'customer_orders_view' : (psycopg2.errors.UndefinedTable) relation "orders" does not exist
LINE 10:  INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.orde...
                     ^
QUERY:  SELECT 
		c.id AS customer_id,
		c.name AS customer_name,
		c.phone_number,
		o.id AS order_id,
		o.name AS order_name,
		o.amount,
		o.order_date
	FROM customers c 
	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC
CONTEXT:  PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY

[SQL: SELECT * FROM customer_orders_view]
(Background on this error at: https://sqlalche.me/e/20/f405)

### Analysis:
```python

---------------------------------------------------------------------------
UndefinedColumn                           Traceback (most recent call last)
File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1967, in Connection._exec_single_context(self, dialect, context, statement, parameters)
   1966     if not evt_handled:
-> 1967         self.dialect.do_execute(
   1968             cursor, str_statement, effective_parameters, context
   1969         )
   1971 if self._has_events or self.engine._has_events:

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/default.py:951, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
    950 def do_execute(self, cursor, statement, parameters, context=None):
--> 951     cursor.execute(statement, parameters)

UndefinedColumn: column o.id does not exist
LINE 5:   o.id AS order_id,
          ^
HINT:  Perhaps you meant to reference the column "c.id".
QUERY:  SELECT 
		c.id AS customer_id,
		c.name AS customer_name,
		c.phone_number,
		o.id AS order_id,
		o.name AS order_name,
		o.amount,
		o.order_date
	FROM customers c 
	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC
CONTEXT:  PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY


The above exception was the direct cause of the following exception:

ProgrammingError                          Traceback (most recent call last)
Cell In[34], line 8, in read_db_table_db_view_support(config)
      7 if not len(config.columns):
----> 8     df = pd.read_sql_query(f"SELECT * FROM {config.database_view_name}",con=config.engine)
      9 else:

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:526, in read_sql_query(sql, con, index_col, coerce_float, params, parse_dates, chunksize, dtype, dtype_backend)
    525 with pandasSQL_builder(con) as pandas_sql:
--> 526     return pandas_sql.read_query(
    527         sql,
    528         index_col=index_col,
    529         params=params,
    530         coerce_float=coerce_float,
    531         parse_dates=parse_dates,
    532         chunksize=chunksize,
    533         dtype=dtype,
    534         dtype_backend=dtype_backend,
    535     )

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1836, in SQLDatabase.read_query(self, sql, index_col, coerce_float, parse_dates, params, chunksize, dtype, dtype_backend)
   1790 """
   1791 Read SQL query into a DataFrame.
   1792 
   (...)
   1834 
   1835 """
-> 1836 result = self.execute(sql, params)
   1837 columns = result.keys()

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1659, in SQLDatabase.execute(self, sql, params)
   1658 if isinstance(sql, str):
-> 1659     return self.con.exec_driver_sql(sql, *args)
   1660 return self.con.execute(sql, *args)

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1779, in Connection.exec_driver_sql(self, statement, parameters, execution_options)
   1778 dialect = self.dialect
-> 1779 ret = self._execute_context(
   1780     dialect,
   1781     dialect.execution_ctx_cls._init_statement,
   1782     statement,
   1783     None,
   1784     execution_options,
   1785     statement,
   1786     distilled_parameters,
   1787 )
   1789 return ret

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1846, in Connection._execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1845 else:
-> 1846     return self._exec_single_context(
   1847         dialect, context, statement, parameters
   1848     )

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1986, in Connection._exec_single_context(self, dialect, context, statement, parameters)
   1985 except BaseException as e:
-> 1986     self._handle_dbapi_exception(
   1987         e, str_statement, effective_parameters, cursor, context
   1988     )
   1990 return result

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2355, in Connection._handle_dbapi_exception(self, e, statement, parameters, cursor, context, is_sub_exec)
   2354     assert sqlalchemy_exception is not None
-> 2355     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
   2356 else:

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1967, in Connection._exec_single_context(self, dialect, context, statement, parameters)
   1966     if not evt_handled:
-> 1967         self.dialect.do_execute(
   1968             cursor, str_statement, effective_parameters, context
   1969         )
   1971 if self._has_events or self.engine._has_events:

File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/default.py:951, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
    950 def do_execute(self, cursor, statement, parameters, context=None):
--> 951     cursor.execute(statement, parameters)

ProgrammingError: (psycopg2.errors.UndefinedColumn) column o.id does not exist
LINE 5:   o.id AS order_id,
          ^
HINT:  Perhaps you meant to reference the column "c.id".
QUERY:  SELECT 
		c.id AS customer_id,
		c.name AS customer_name,
		c.phone_number,
		o.id AS order_id,
		o.name AS order_name,
		o.amount,
		o.order_date
	FROM customers c 
	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC
CONTEXT:  PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY

[SQL: SELECT * FROM customer_orders_view]
(Background on this error at: https://sqlalche.me/e/20/f405)

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
Cell In[52], line 3
      1 db_view_name = "customer_orders_view"
      2 database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
----> 3 database_view_df = read_db_table_db_view_support(database_view_table)
      4 database_view_df

Cell In[34], line 13, in read_db_table_db_view_support(config)
     11         return df
     12     except Exception as e:
---> 13         raise RuntimeError(f"Failed to read table '{config.database_view_name}' : {e}")
     14 if config.table_name and not config.database_view_name:
     15     try:

RuntimeError: Failed to read table 'customer_orders_view' : (psycopg2.errors.UndefinedColumn) column o.id does not exist
LINE 5:   o.id AS order_id,
          ^
HINT:  Perhaps you meant to reference the column "c.id".
QUERY:  SELECT 
		c.id AS customer_id,
		c.name AS customer_name,
		c.phone_number,
		o.id AS order_id,
		o.name AS order_name,
		o.amount,
		o.order_date
	FROM customers c 
	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC
CONTEXT:  PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY

[SQL: SELECT * FROM customer_orders_view]
(Background on this error at: https://sqlalche.me/e/20/f405)

```
<br>

As I was experimenting with the function ```write_df_to_database``` what writes the dataFrame to the database and the function ```read_db_table_db_view_support``` that uses the database view to generate a dataFrame in pandas. I was blasted by this error.

#### When this error happened?
I step foot on this devastating landmine when ->
- I droped the table from the postgreSQL database using PGAdmin and then Tried to write the dataFrame created via csv file into the database.
- Then I fired up this function ```read_db_table_db_view_support``` to get the dataFrame using a database view.

#### Why is this error blew up in my face?
Reasons :
- If you look closely in this function ```write_df_to_database``` you will see that there is a code that removes id column from the dataFrame
  ```python
    if "id" in config.df.columns:
        config.df = config.df.drop(columns=["id"])
  ```
  - Reason for implementing this code:
    - I wrote this code to remove the id column from the dataFrame because posgres was complaining saying the ids (primary key) was not unique
    - Reason for this is because the id (representing primary keys) in the dataFrame and the actual primary keys in the database were cause the unique constraint to fail in case where there exist some data in the table in which case the pandas method ```to_sql()``` will use the ```if_exists="append"``` argument to append new data into the existing table.
  - Now the same code is cause problem in case when the table does not exist in the database in which case the pandas method ```to_sql()``` will create a new table and then insert the data from the dataFrame into the newly created table but since this code removed the id column from the dataFrame the data inserted in the table also doesn't have the id column in it.
  - This is causes the database View I wrote to fail because it uses inner join to join the two tables and return a view back to pandas hence I am seeing this error.

#### Solution to this problem :
- There is a package in sqlalchemy that I can use called ```inspect``` to check if the table name passed in the function ```write_df_to_database``` exists in the database or not. 
- If the table is there in the database then remove the id column from the dataFrame.
- If the table is not present in the database then don't remove the id column from the dataFrame.

### Fixing ```write_df_to_database``` function to make sure it don't fail when there is no table present in the database
- Fixing the logic to remove id column from the database because its causing the this function ```read_db_table_db_view_support``` to fail when table does not exist in the database.
- This function works fine when the table is already present and new data is to be appended in the existing table.

In [51]:
class WritePandasToDBModel(BaseModel):
    df:pd.DataFrame=Field(default=None)
    table_name:str=Field(default=None)
    index:bool=Field(default=False)
    if_exists:str=Field(default=False)
    engine:Engine
    model_config={
        "arbitrary_types_allowed":True
    }
def write_df_to_database(config:WritePandasToDBModel):
    if config.df is None or config.df.empty:
        raise ValueError("DataFrame is required and cannot be empty to write into the database")
    if not table_name:
        raise ValueError("Table name is required to be able to write the data into the database.")
    if config.table_name:
        try:
            # remove the id (primary key column) if exists after checking if the table name passed in the function exists in the database.
            inspector = inspect(config.engine)
            if inspector.has_table(config.table_name) and "id" in config.df.columns:
                config.df = config.df.drop(columns=["id"])
            config.df.to_sql(name=config.table_name,con=config.engine,index=config.index,if_exists=config.if_exists)
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}':{e}")

#### I have droped the orders table from the database using PGAdmin Let's see if this updated function fixes the issue

**Create the orders table in the database and then insert data into it:**

In [52]:
table_name = "orders"
engine=engine_obj
if_exists="append"
index=False
dataFrame=csv_orders_df
write_to_db = WritePandasToDBModel(df=dataFrame,table_name=table_name,engine=engine,if_exists=if_exists,index=index)
write_df_to_database =write_df_to_database(write_to_db)

**Now lets use the database view to generate a dataFrame in pandas:**

In [53]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

RuntimeError: Failed to read table 'customer_orders_view' : (psycopg2.errors.DatatypeMismatch) structure of query does not match function result type
DETAIL:  Returned type text does not match expected type character varying in column 5.
CONTEXT:  SQL statement "SELECT 
		c.id AS customer_id,
		c.name AS customer_name,
		c.phone_number,
		o.id AS order_id,
		o.name AS order_name,
		o.amount,
		o.order_date
	FROM customers c 
	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC"
PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY

[SQL: SELECT * FROM customer_orders_view]
(Background on this error at: https://sqlalche.me/e/20/f405)

### Analysis
After making changes to the function ```write_df_to_database``` I am getting a new type of error 
```python
    ---------------------------------------------------------------------------
    DatatypeMismatch                          Traceback (most recent call last)
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1967, in Connection._exec_single_context(self, dialect, context, statement, parameters)
       1966     if not evt_handled:
    -> 1967         self.dialect.do_execute(
       1968             cursor, str_statement, effective_parameters, context
       1969         )
       1971 if self._has_events or self.engine._has_events:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/default.py:951, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
        950 def do_execute(self, cursor, statement, parameters, context=None):
    --> 951     cursor.execute(statement, parameters)
    
    DatatypeMismatch: structure of query does not match function result type
    DETAIL:  Returned type text does not match expected type character varying in column 5.
    CONTEXT:  SQL statement "SELECT 
    		c.id AS customer_id,
    		c.name AS customer_name,
    		c.phone_number,
    		o.id AS order_id,
    		o.name AS order_name,
    		o.amount,
    		o.order_date
    	FROM customers c 
    	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC"
    PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY
    
    
    The above exception was the direct cause of the following exception:
    
    ProgrammingError                          Traceback (most recent call last)
    Cell In[34], line 8, in read_db_table_db_view_support(config)
          7 if not len(config.columns):
    ----> 8     df = pd.read_sql_query(f"SELECT * FROM {config.database_view_name}",con=config.engine)
          9 else:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:526, in read_sql_query(sql, con, index_col, coerce_float, params, parse_dates, chunksize, dtype, dtype_backend)
        525 with pandasSQL_builder(con) as pandas_sql:
    --> 526     return pandas_sql.read_query(
        527         sql,
        528         index_col=index_col,
        529         params=params,
        530         coerce_float=coerce_float,
        531         parse_dates=parse_dates,
        532         chunksize=chunksize,
        533         dtype=dtype,
        534         dtype_backend=dtype_backend,
        535     )
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1836, in SQLDatabase.read_query(self, sql, index_col, coerce_float, parse_dates, params, chunksize, dtype, dtype_backend)
       1790 """
       1791 Read SQL query into a DataFrame.
       1792 
       (...)
       1834 
       1835 """
    -> 1836 result = self.execute(sql, params)
       1837 columns = result.keys()
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1659, in SQLDatabase.execute(self, sql, params)
       1658 if isinstance(sql, str):
    -> 1659     return self.con.exec_driver_sql(sql, *args)
       1660 return self.con.execute(sql, *args)
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1779, in Connection.exec_driver_sql(self, statement, parameters, execution_options)
       1778 dialect = self.dialect
    -> 1779 ret = self._execute_context(
       1780     dialect,
       1781     dialect.execution_ctx_cls._init_statement,
       1782     statement,
       1783     None,
       1784     execution_options,
       1785     statement,
       1786     distilled_parameters,
       1787 )
       1789 return ret
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1846, in Connection._execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
       1845 else:
    -> 1846     return self._exec_single_context(
       1847         dialect, context, statement, parameters
       1848     )
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1986, in Connection._exec_single_context(self, dialect, context, statement, parameters)
       1985 except BaseException as e:
    -> 1986     self._handle_dbapi_exception(
       1987         e, str_statement, effective_parameters, cursor, context
       1988     )
       1990 return result
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2355, in Connection._handle_dbapi_exception(self, e, statement, parameters, cursor, context, is_sub_exec)
       2354     assert sqlalchemy_exception is not None
    -> 2355     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
       2356 else:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1967, in Connection._exec_single_context(self, dialect, context, statement, parameters)
       1966     if not evt_handled:
    -> 1967         self.dialect.do_execute(
       1968             cursor, str_statement, effective_parameters, context
       1969         )
       1971 if self._has_events or self.engine._has_events:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/default.py:951, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
        950 def do_execute(self, cursor, statement, parameters, context=None):
    --> 951     cursor.execute(statement, parameters)
    
    ProgrammingError: (psycopg2.errors.DatatypeMismatch) structure of query does not match function result type
    DETAIL:  Returned type text does not match expected type character varying in column 5.
    CONTEXT:  SQL statement "SELECT 
    		c.id AS customer_id,
    		c.name AS customer_name,
    		c.phone_number,
    		o.id AS order_id,
    		o.name AS order_name,
    		o.amount,
    		o.order_date
    	FROM customers c 
    	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC"
    PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY
    
    [SQL: SELECT * FROM customer_orders_view]
    (Background on this error at: https://sqlalche.me/e/20/f405)
    
    During handling of the above exception, another exception occurred:
    
    RuntimeError                              Traceback (most recent call last)
    Cell In[60], line 3
          1 db_view_name = "customer_orders_view"
          2 database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
    ----> 3 database_view_df = read_db_table_db_view_support(database_view_table)
          4 database_view_df
    
    Cell In[34], line 13, in read_db_table_db_view_support(config)
         11         return df
         12     except Exception as e:
    ---> 13         raise RuntimeError(f"Failed to read table '{config.database_view_name}' : {e}")
         14 if config.table_name and not config.database_view_name:
         15     try:
    
    RuntimeError: Failed to read table 'customer_orders_view' : (psycopg2.errors.DatatypeMismatch) structure of query does not match function result type
    DETAIL:  Returned type text does not match expected type character varying in column 5.
    CONTEXT:  SQL statement "SELECT 
    		c.id AS customer_id,
    		c.name AS customer_name,
    		c.phone_number,
    		o.id AS order_id,
    		o.name AS order_name,
    		o.amount,
    		o.order_date
    	FROM customers c 
    	INNER JOIN Orders o on c.id = o.customer_id ORDER BY o.order_date DESC"
    PL/pgSQL function get_customer_orders() line 4 at RETURN QUERY
    
    [SQL: SELECT * FROM customer_orders_view]
    (Background on this error at: https://sqlalche.me/e/20/f405)
```

<br>

#### What is this error?
This error means my database function declared the return type to be (```VARCHAR```) does not exactly match the column type in your actual ```orders``` table after recreation.

#### Why it worked before dropping the table?
- Originally , my ```ordrs.name``` column was defined as ```VARCHAR(255)```.
- My database function ```get_customer_orders()``` expected ```order_name VARCHAR```.
- ```VARCHAR``` -> ```VARCHAR(255)``` cast works fine so no error.

#### Why it fails after dropping and re-creating?
When I re-created ```orders``` via Pandas ```to_sql```, SQLAlchemy doesn't use your original DDL (with ```VARCHAR(255)```), it maps pandas ```object``` dtype to TEXT by default. <br>
So now:
- ```orders.name``` column is ```TEXT```
- My function still declares ```order_name VARCHAR```
- PosgreSQL complains: ```text``` $\neq$ ```varchar``` (strict type mismatch inside ```RETURNS TABLE```).
- That's the error which I am seeing here.

#### Solutions:
There are two solutions to this problem
- Solution 1: Make the database function to return ```TEXT``` instead of ```VARCHAR``` Since Pandas -> PostgreSQL defaults to ```TEXT```, just declare it explicitly.
- Solution 2: Force Pandas ```to_sql()``` to preserve schema

<br>

Here in my case I am going with the solution 2 : Where I am going to force pandas ```to_sql()``` method to preserve posgreSQL schema.

#### Fixing the function
Here I am going to force the ```to_sql()``` method to follow the shcema of postgreSQL orders Table so that when the orders table is dropped using PGAdmin and ```to_sql()``` method creates a new table in the database.

In [54]:
from sqlalchemy import String, Numeric, DateTime, BigInteger, Column, Table, MetaData

metadata = MetaData()

orders = Table(
    "orders",metadata,
    Column("id",BigInteger,primary_key=True),
    Column("customer_id",BigInteger,nullable=False),
    Column("name",String(255),nullable=False),
    Column("amount",Numeric(12,2),nullable=False),
    Column("order_date",DateTime,nullable=False),
)

customers = Table(
    "customers",metadata,
    Column("id",BigInteger,primary_key=True),
    Column("name",String(255),nullable=False),
    Column("phone_number",String(10),nullable=False)
)

class WritePandasToDBModel(BaseModel):
    df:pd.DataFrame=Field(default=None)
    table_name:str=Field(default=None)
    index:bool=Field(default=False)
    if_exists:str=Field(default=False)
    engine:Engine
    model_config={
        "arbitrary_types_allowed":True
    }
    
def write_df_to_database(config:WritePandasToDBModel):
    if config.df is None or config.df.empty:
        raise ValueError("DataFrame is required and cannot be empty to write into the database")
    if not table_name:
        raise ValueError("Table name is required to be able to write the data into the database.")
    if config.table_name:
        try:
            # Creates the table if it does not exist in the database
            inspector = inspect(config.engine)
            if not inspector.has_table(config.table_name):
                metadata.create_all(engine) 
            # remove the id (primary key column) if exists
            if "id" in config.df.columns:
                config.df = config.df.drop(columns=["id"])
            config.df.to_sql(name=config.table_name,con=config.engine,index=config.index,if_exists=config.if_exists)
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}':{e}")

In [55]:
table_name = "orders"
engine=engine_obj
if_exists="append"
index=False
dataFrame=csv_orders_df
write_to_db = WritePandasToDBModel(df=dataFrame,table_name=table_name,engine=engine,if_exists=if_exists,index=index)
write_df_to_database =write_df_to_database(write_to_db)

RuntimeError: Failed to read table 'orders':(psycopg2.errors.InvalidTextRepresentation) invalid input syntax for type numeric: "$9,978.12"
LINE 1: ...r_id, name, amount, order_date) VALUES (1, 'Axe', '$9,978.12...
                                                             ^

[SQL: INSERT INTO orders (customer_id, name, amount, order_date) VALUES (%(customer_id__0)s, %(name__0)s, %(amount__0)s, %(order_date__0)s), (%(customer_id__1)s, %(name__1)s, %(amount__1)s, %(order_date__1)s), (%(customer_id__2)s, %(name__2)s, %(amount__2) ... 696 characters truncated ... __12)s, %(order_date__12)s), (%(customer_id__13)s, %(name__13)s, %(amount__13)s, %(order_date__13)s)]
[parameters: {'customer_id__0': 1, 'name__0': 'Axe', 'order_date__0': '2025-09-01 12:01:26.625465', 'amount__0': '$9,978.12', 'customer_id__1': 2, 'name__1': 'Sledge Hammer', 'order_date__1': '2025-09-01 12:01:26.625465', 'amount__1': '$1,000.12', 'customer_id__2': 3, 'name__2': 'Showel', 'order_date__2': '2025-09-01 12:01:26.625465', 'amount__2': '$2,470.12', 'customer_id__3': 4, 'name__3': 'Cross blade Axe', 'order_date__3': '2025-09-01 12:01:26.625465', 'amount__3': '$7,620.12', 'customer_id__4': 5, 'name__4': 'Pick Axe', 'order_date__4': '2025-09-01 12:01:26.625465', 'amount__4': '$1,170.12', 'customer_id__5': 6, 'name__5': 'Michetti', 'order_date__5': '2025-09-01 12:01:26.625465', 'amount__5': '$1,500.12', 'customer_id__6': 7, 'name__6': 'Crow-bar', 'order_date__6': '2025-09-01 12:01:26.625465', 'amount__6': '$1,850.12', 'customer_id__7': 8, 'name__7': 'Titanium base ball bat', 'order_date__7': '2025-09-01 12:01:26.625465', 'amount__7': '$3,650.12', 'customer_id__8': 9, 'name__8': 'Mace', 'order_date__8': '2025-09-01 12:01:26.625465', 'amount__8': '$3,650.12', 'customer_id__9': 10, 'name__9': 'Gurkha dagger', 'order_date__9': '2025-09-01 12:01:26.625465', 'amount__9': '$6,000.12', 'customer_id__10': 11, 'name__10': 'Throwing knife', 'order_date__10': '2025-09-01 12:01:26.625465', 'amount__10': '$3,600.12', 'customer_id__11': 12, 'name__11': 'Tawma Hawk Axe', 'order_date__11': '2025-09-01 12:01:26.625465', 'amount__11': '$7,470.12', 'customer_id__12': 13, 'name__12': 'ice Axe', 'order_date__12': '2025-09-01 12:01:26.625465', 'amount__12': '$8,000.12', 'customer_id__13': 14, 'name__13': 'Military knife', 'order_date__13': '2025-09-01 12:01:26.625465', 'amount__13': '$9,110.12'}]
(Background on this error at: https://sqlalche.me/e/20/9h9h)

### Analysis:
```python
    ---------------------------------------------------------------------------
    InvalidTextRepresentation                 Traceback (most recent call last)
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2118, in Connection._exec_insertmany_context(self, dialect, context)
       2117     else:
    -> 2118         dialect.do_execute(
       2119             cursor,
       2120             sub_stmt,
       2121             sub_params,
       2122             context,
       2123         )
       2125 except BaseException as e:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/default.py:951, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
        950 def do_execute(self, cursor, statement, parameters, context=None):
    --> 951     cursor.execute(statement, parameters)
    
    InvalidTextRepresentation: invalid input syntax for type numeric: "$9,978.12"
    LINE 1: ...r_id, name, amount, order_date) VALUES (1, 'Axe', '$9,978.12...
                                                                 ^
    
    
    The above exception was the direct cause of the following exception:
    
    DataError                                 Traceback (most recent call last)
    Cell In[54], line 45, in write_df_to_database(config)
         44         config.df = config.df.drop(columns=["id"])
    ---> 45     config.df.to_sql(name=config.table_name,con=config.engine,index=config.index,if_exists=config.if_exists)
         46 except Exception as e:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/util/_decorators.py:333, in deprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
        328     warnings.warn(
        329         msg.format(arguments=_format_argument_list(allow_args)),
        330         FutureWarning,
        331         stacklevel=find_stack_level(),
        332     )
    --> 333 return func(*args, **kwargs)
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/core/generic.py:3087, in NDFrame.to_sql(self, name, con, schema, if_exists, index, index_label, chunksize, dtype, method)
       3085 from pandas.io import sql
    -> 3087 return sql.to_sql(
       3088     self,
       3089     name,
       3090     con,
       3091     schema=schema,
       3092     if_exists=if_exists,
       3093     index=index,
       3094     index_label=index_label,
       3095     chunksize=chunksize,
       3096     dtype=dtype,
       3097     method=method,
       3098 )
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:842, in to_sql(frame, name, con, schema, if_exists, index, index_label, chunksize, dtype, method, engine, **engine_kwargs)
        841 with pandasSQL_builder(con, schema=schema, need_transaction=True) as pandas_sql:
    --> 842     return pandas_sql.to_sql(
        843         frame,
        844         name,
        845         if_exists=if_exists,
        846         index=index,
        847         index_label=index_label,
        848         schema=schema,
        849         chunksize=chunksize,
        850         dtype=dtype,
        851         method=method,
        852         engine=engine,
        853         **engine_kwargs,
        854     )
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:2018, in SQLDatabase.to_sql(self, frame, name, if_exists, index, index_label, schema, chunksize, dtype, method, engine, **engine_kwargs)
       2008 table = self.prep_table(
       2009     frame=frame,
       2010     name=name,
       (...)
       2015     dtype=dtype,
       2016 )
    -> 2018 total_inserted = sql_engine.insert_records(
       2019     table=table,
       2020     con=self.con,
       2021     frame=frame,
       2022     name=name,
       2023     index=index,
       2024     schema=schema,
       2025     chunksize=chunksize,
       2026     method=method,
       2027     **engine_kwargs,
       2028 )
       2030 self.check_case_sensitive(name=name, schema=schema)
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1567, in SQLAlchemyEngine.insert_records(self, table, con, frame, name, index, schema, chunksize, method, **engine_kwargs)
       1566     raise ValueError("inf cannot be used with MySQL") from err
    -> 1567 raise err
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1558, in SQLAlchemyEngine.insert_records(self, table, con, frame, name, index, schema, chunksize, method, **engine_kwargs)
       1557 try:
    -> 1558     return table.insert(chunksize=chunksize, method=method)
       1559 except exc.StatementError as err:
       1560     # GH34431
       1561     # https://stackoverflow.com/a/67358288/6067848
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1119, in SQLTable.insert(self, chunksize, method)
       1118 chunk_iter = zip(*(arr[start_i:end_i] for arr in data_list))
    -> 1119 num_inserted = exec_insert(conn, keys, chunk_iter)
       1120 # GH 46891
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/pandas/io/sql.py:1010, in SQLTable._execute_insert(self, conn, keys, data_iter)
       1009 data = [dict(zip(keys, row)) for row in data_iter]
    -> 1010 result = conn.execute(self.table.insert(), data)
       1011 return result.rowcount
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1419, in Connection.execute(self, statement, parameters, execution_options)
       1418 else:
    -> 1419     return meth(
       1420         self,
       1421         distilled_parameters,
       1422         execution_options or NO_OPTIONS,
       1423     )
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/sql/elements.py:526, in ClauseElement._execute_on_connection(self, connection, distilled_params, execution_options)
        525         assert isinstance(self, Executable)
    --> 526     return connection._execute_clauseelement(
        527         self, distilled_params, execution_options
        528     )
        529 else:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1641, in Connection._execute_clauseelement(self, elem, distilled_parameters, execution_options)
       1633 compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
       1634     dialect=dialect,
       1635     compiled_cache=compiled_cache,
       (...)
       1639     linting=self.dialect.compiler_linting | compiler.WARN_LINTING,
       1640 )
    -> 1641 ret = self._execute_context(
       1642     dialect,
       1643     dialect.execution_ctx_cls._init_compiled,
       1644     compiled_sql,
       1645     distilled_parameters,
       1646     execution_options,
       1647     compiled_sql,
       1648     distilled_parameters,
       1649     elem,
       1650     extracted_params,
       1651     cache_hit=cache_hit,
       1652 )
       1653 if has_events:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1844, in Connection._execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
       1843 if context.execute_style is ExecuteStyle.INSERTMANYVALUES:
    -> 1844     return self._exec_insertmany_context(dialect, context)
       1845 else:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2126, in Connection._exec_insertmany_context(self, dialect, context)
       2125 except BaseException as e:
    -> 2126     self._handle_dbapi_exception(
       2127         e,
       2128         sql_util._long_statement(sub_stmt),
       2129         sub_params,
       2130         cursor,
       2131         context,
       2132         is_sub_exec=True,
       2133     )
       2135 if engine_events:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2355, in Connection._handle_dbapi_exception(self, e, statement, parameters, cursor, context, is_sub_exec)
       2354     assert sqlalchemy_exception is not None
    -> 2355     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
       2356 else:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2118, in Connection._exec_insertmany_context(self, dialect, context)
       2117     else:
    -> 2118         dialect.do_execute(
       2119             cursor,
       2120             sub_stmt,
       2121             sub_params,
       2122             context,
       2123         )
       2125 except BaseException as e:
    
    File ~/miniconda3/envs/tensorflow/lib/python3.10/site-packages/sqlalchemy/engine/default.py:951, in DefaultDialect.do_execute(self, cursor, statement, parameters, context)
        950 def do_execute(self, cursor, statement, parameters, context=None):
    --> 951     cursor.execute(statement, parameters)
    
    DataError: (psycopg2.errors.InvalidTextRepresentation) invalid input syntax for type numeric: "$9,978.12"
    LINE 1: ...r_id, name, amount, order_date) VALUES (1, 'Axe', '$9,978.12...
                                                                 ^
    
    [SQL: INSERT INTO orders (customer_id, name, amount, order_date) VALUES (%(customer_id__0)s, %(name__0)s, %(amount__0)s, %(order_date__0)s), (%(customer_id__1)s, %(name__1)s, %(amount__1)s, %(order_date__1)s), (%(customer_id__2)s, %(name__2)s, %(amount__2) ... 696 characters truncated ... __12)s, %(order_date__12)s), (%(customer_id__13)s, %(name__13)s, %(amount__13)s, %(order_date__13)s)]
    [parameters: {'customer_id__0': 1, 'name__0': 'Axe', 'order_date__0': '2025-09-01 12:01:26.625465', 'amount__0': '$9,978.12', 'customer_id__1': 2, 'name__1': 'Sledge Hammer', 'order_date__1': '2025-09-01 12:01:26.625465', 'amount__1': '$1,000.12', 'customer_id__2': 3, 'name__2': 'Showel', 'order_date__2': '2025-09-01 12:01:26.625465', 'amount__2': '$2,470.12', 'customer_id__3': 4, 'name__3': 'Cross blade Axe', 'order_date__3': '2025-09-01 12:01:26.625465', 'amount__3': '$7,620.12', 'customer_id__4': 5, 'name__4': 'Pick Axe', 'order_date__4': '2025-09-01 12:01:26.625465', 'amount__4': '$1,170.12', 'customer_id__5': 6, 'name__5': 'Michetti', 'order_date__5': '2025-09-01 12:01:26.625465', 'amount__5': '$1,500.12', 'customer_id__6': 7, 'name__6': 'Crow-bar', 'order_date__6': '2025-09-01 12:01:26.625465', 'amount__6': '$1,850.12', 'customer_id__7': 8, 'name__7': 'Titanium base ball bat', 'order_date__7': '2025-09-01 12:01:26.625465', 'amount__7': '$3,650.12', 'customer_id__8': 9, 'name__8': 'Mace', 'order_date__8': '2025-09-01 12:01:26.625465', 'amount__8': '$3,650.12', 'customer_id__9': 10, 'name__9': 'Gurkha dagger', 'order_date__9': '2025-09-01 12:01:26.625465', 'amount__9': '$6,000.12', 'customer_id__10': 11, 'name__10': 'Throwing knife', 'order_date__10': '2025-09-01 12:01:26.625465', 'amount__10': '$3,600.12', 'customer_id__11': 12, 'name__11': 'Tawma Hawk Axe', 'order_date__11': '2025-09-01 12:01:26.625465', 'amount__11': '$7,470.12', 'customer_id__12': 13, 'name__12': 'ice Axe', 'order_date__12': '2025-09-01 12:01:26.625465', 'amount__12': '$8,000.12', 'customer_id__13': 14, 'name__13': 'Military knife', 'order_date__13': '2025-09-01 12:01:26.625465', 'amount__13': '$9,110.12'}]
    (Background on this error at: https://sqlalche.me/e/20/9h9h)
    
    During handling of the above exception, another exception occurred:
    
    RuntimeError                              Traceback (most recent call last)
    Cell In[55], line 7
          5 dataFrame=csv_orders_df
          6 write_to_db = WritePandasToDBModel(df=dataFrame,table_name=table_name,engine=engine,if_exists=if_exists,index=index)
    ----> 7 write_df_to_database =write_df_to_database(write_to_db)
    
    Cell In[54], line 47, in write_df_to_database(config)
         45     config.df.to_sql(name=config.table_name,con=config.engine,index=config.index,if_exists=config.if_exists)
         46 except Exception as e:
    ---> 47     raise RuntimeError(f"Failed to read table '{config.table_name}':{e}")
    
    RuntimeError: Failed to read table 'orders':(psycopg2.errors.InvalidTextRepresentation) invalid input syntax for type numeric: "$9,978.12"
    LINE 1: ...r_id, name, amount, order_date) VALUES (1, 'Axe', '$9,978.12...
                                                                 ^
    
    [SQL: INSERT INTO orders (customer_id, name, amount, order_date) VALUES (%(customer_id__0)s, %(name__0)s, %(amount__0)s, %(order_date__0)s), (%(customer_id__1)s, %(name__1)s, %(amount__1)s, %(order_date__1)s), (%(customer_id__2)s, %(name__2)s, %(amount__2) ... 696 characters truncated ... __12)s, %(order_date__12)s), (%(customer_id__13)s, %(name__13)s, %(amount__13)s, %(order_date__13)s)]
    [parameters: {'customer_id__0': 1, 'name__0': 'Axe', 'order_date__0': '2025-09-01 12:01:26.625465', 'amount__0': '$9,978.12', 'customer_id__1': 2, 'name__1': 'Sledge Hammer', 'order_date__1': '2025-09-01 12:01:26.625465', 'amount__1': '$1,000.12', 'customer_id__2': 3, 'name__2': 'Showel', 'order_date__2': '2025-09-01 12:01:26.625465', 'amount__2': '$2,470.12', 'customer_id__3': 4, 'name__3': 'Cross blade Axe', 'order_date__3': '2025-09-01 12:01:26.625465', 'amount__3': '$7,620.12', 'customer_id__4': 5, 'name__4': 'Pick Axe', 'order_date__4': '2025-09-01 12:01:26.625465', 'amount__4': '$1,170.12', 'customer_id__5': 6, 'name__5': 'Michetti', 'order_date__5': '2025-09-01 12:01:26.625465', 'amount__5': '$1,500.12', 'customer_id__6': 7, 'name__6': 'Crow-bar', 'order_date__6': '2025-09-01 12:01:26.625465', 'amount__6': '$1,850.12', 'customer_id__7': 8, 'name__7': 'Titanium base ball bat', 'order_date__7': '2025-09-01 12:01:26.625465', 'amount__7': '$3,650.12', 'customer_id__8': 9, 'name__8': 'Mace', 'order_date__8': '2025-09-01 12:01:26.625465', 'amount__8': '$3,650.12', 'customer_id__9': 10, 'name__9': 'Gurkha dagger', 'order_date__9': '2025-09-01 12:01:26.625465', 'amount__9': '$6,000.12', 'customer_id__10': 11, 'name__10': 'Throwing knife', 'order_date__10': '2025-09-01 12:01:26.625465', 'amount__10': '$3,600.12', 'customer_id__11': 12, 'name__11': 'Tawma Hawk Axe', 'order_date__11': '2025-09-01 12:01:26.625465', 'amount__11': '$7,470.12', 'customer_id__12': 13, 'name__12': 'ice Axe', 'order_date__12': '2025-09-01 12:01:26.625465', 'amount__12': '$8,000.12', 'customer_id__13': 14, 'name__13': 'Military knife', 'order_date__13': '2025-09-01 12:01:26.625465', 'amount__13': '$9,110.12'}]
    (Background on this error at: https://sqlalche.me/e/20/9h9h)
```

<br>

Focus on this line 

```python
    RuntimeError: Failed to read table 'orders':(psycopg2.errors.InvalidTextRepresentation) invalid input syntax for type numeric: "$9,978.12"
    LINE 1: ...r_id, name, amount, order_date) VALUES (1, 'Axe', '$9,978.12...
```

<br>

#### Why is error happening?
The error is happening because your amount column in the CSV (and therefore your DataFrame) has currency-formatted strings like "$9,978.12", but in PostgreSQL the column is defined as Numeric(12,2) (expects a pure number).

So when Pandas’ to_sql tries to insert "$9,978.12" into a NUMERIC column, PostgreSQL rejects it

#### Solution:
I have to add a clean up logic in the ```write_df_to_database()``` function for the dataFrame

In [63]:
from sqlalchemy import String, Numeric, DateTime, BigInteger, Column, Table, MetaData

metadata = MetaData()

orders = Table(
    "orders",metadata,
    Column("id",BigInteger,primary_key=True),
    Column("customer_id",BigInteger,nullable=False),
    Column("name",String(255),nullable=False),
    Column("amount",Numeric(12,2),nullable=False),
    Column("order_date",DateTime,nullable=False),
)

customers = Table(
    "customers",metadata,
    Column("id",BigInteger,primary_key=True),
    Column("name",String(255),nullable=False),
    Column("phone_number",String(10),nullable=False)
)

class WritePandasToDBModel(BaseModel):
    df:pd.DataFrame=Field(default=None)
    table_name:str=Field(default=None)
    index:bool=Field(default=False)
    if_exists:str=Field(default=False)
    engine:Engine
    model_config={
        "arbitrary_types_allowed":True
    }
    
def write_df_to_database(config:WritePandasToDBModel):
    if config.df is None or config.df.empty:
        raise ValueError("DataFrame is required and cannot be empty to write into the database")
    if not table_name:
        raise ValueError("Table name is required to be able to write the data into the database.")
    if config.table_name:
        try:
            # Creates the table if it does not exist in the database
            inspector = inspect(config.engine)
            if not inspector.has_table(config.table_name):
                metadata.create_all(engine) 
            # remove the id (primary key column) if exists
            if "id" in config.df.columns:
                config.df = config.df.drop(columns=["id"])
            # claen up the orders dataFrame before feeding it to database for storage
            if "amount" in config.df.columns:
                config.df["amount"] = (
                    config.df["amount"].replace(r"[\$,]","",regex=True) # remove the $ and commas from the amount column
                    .astype(float)
                )
            config.df.to_sql(name=config.table_name,con=config.engine,index=config.index,if_exists=config.if_exists)
        except Exception as e:
            raise RuntimeError(f"Failed to read table '{config.table_name}':{e}")

In [64]:
table_name = "orders"
engine=engine_obj
if_exists="append"
index=False
dataFrame=csv_orders_df
write_to_db = WritePandasToDBModel(df=dataFrame,table_name=table_name,engine=engine,if_exists=if_exists,index=index)
write_df_to_database =write_df_to_database(write_to_db)

#### Now lets use the database view to generate a dataFrame in pandas:

In [66]:
db_view_name = "customer_orders_view"
database_view_table = ReadDBViewModel(database_view_name=db_view_name, engine=engine_obj)
database_view_df = read_db_table_db_view_support(database_view_table)
database_view_df

,customer_id,customer_name,phone_number,order_id,order_name,amount,order_date
0,1,Rollex,9972688590,1,Axe,9978.12,2025-09-01 12:01:26.625465
1,2,Ballistic,1000688590,2,Sledge Hammer,1000.12,2025-09-01 12:01:26.625465
2,3,Barricade,2472688590,3,Showel,2470.12,2025-09-01 12:01:26.625465
3,4,Bullet,7622688590,4,Cross blade Axe,7620.12,2025-09-01 12:01:26.625465
4,5,Barbatos,1172688590,5,Pick Axe,1170.12,2025-09-01 12:01:26.625465
5,6,Artillery,1502688590,6,Michetti,1500.12,2025-09-01 12:01:26.625465
6,7,Shotgun,1852688590,7,Crow-bar,1850.12,2025-09-01 12:01:26.625465
7,8,Kombat,3652688590,8,Titanium base ball bat,3650.12,2025-09-01 12:01:26.625465
8,9,Lockdown,1442688590,9,Mace,3650.12,2025-09-01 12:01:26.625465
9,10,Ignition,6000688590,10,Gurkha dagger,6000.12,2025-09-01 12:01:26.625465
